File: 1172-dinner-plate-stacks.ts
class DinnerPlates {
  private stacks: number[][] = [];

  constructor(private readonly capacity: number) {}

  push(val: number): void {
    const index = this.findLeftmostIndex();

    this.stacks[index].push(val);
  }

  pop(): number {
    const index = this.findRightmostIndex();
    if (index === -1) {
      return -1;
    }

    return this.popAtStack(index);
  }

  popAtStack(index: number): number {
    const val = this.stacks[index].pop();

    if (typeof val === 'undefined') {
      return -1;
    }

    return val;
  }

  private findLeftmostIndex(): number {
    const index = this.stacks.findIndex((x) => x.length < this.capacity);

    if (index !== -1) {
      return index;
    }

    return this.stacks.push([]) - 1;
  }

  private findRightmostIndex(): number {
    for (let i = this.stacks.length - 1; i >= 0; i--) {
      if (this.stacks[i].length > 0) {
        return i;
      }
    }

    return -1;
  }
}

// 2
// 1
// 0 1 2 3 4 ...
const obj = new DinnerPlates(2);
obj.push(1);
obj.push(2);
obj.push(3);

const shouldBe3 = obj.pop();
const shouldBeMinus1 = obj.popAtStack(1);

/*



*/

// obj.push(2);
// obj.push(3);
// obj.push(4);
// var param_2 = obj.pop();
// var param_3 = obj.popAtStack(0);

File: 1172-dinner-plate-stacks.cs
public class DinnerPlates
{
  private int _capacity;
  private SortedSet<int> _pushingSet = new SortedSet<int>();
  private SortedSet<int> _poppingSet = new SortedSet<int>();
  private List<Stack<int>> _stacks = new List<Stack<int>>();

  public DinnerPlates(int capacity)
  {
    _capacity = capacity;
  }

  public void Push(int val)
  {
    var index = _pushingSet.Count == 0 ? _stacks.Count : _pushingSet.Min;

    var stack = GetOrCreateStackByIndex(index);

    stack.Push(val);

    if (stack.Count == _capacity)
    {
      _pushingSet.Remove(index);
    }
    else
    {
      _pushingSet.Add(index);
    }

    _poppingSet.Add(index);
  }

  public int Pop()
  {
    if (_poppingSet.Count == 0)
    {
      return -1;
    }

    var index = _poppingSet.Max;

    return PopAtStack(index);
  }

  public int PopAtStack(int index)
  {
    if (index >= _stacks.Count)
    {
      return -1;
    }

    var stack = _stacks[index];

    if (!stack.TryPop(out int value))
    {
      return -1;
    }

    if (stack.Count == 0)
    {
      _poppingSet.Remove(index);
    }

    _pushingSet.Add(index);

    return value;
  }

  private Stack<int> GetOrCreateStackByIndex(int index)
  {
    var count = _stacks.Count;

    if (index < count)
    {
      return _stacks[index];
    }

    _stacks.AddRange(Enumerable.Range(0, index - count + 1).Select(x => new Stack<int>()));

    return _stacks[index];
  }
}
